src/random_ray/iteration.cpp

```diff
@@ -0,0 +1,610 @@
+
+    // Master rank will gather results and pick valid positions
+    if (mpi::master) {
+      // Initialize temporary vector for receiving positions
+      std::vector<std::vector<Position>> all_position;
+      all_position.resize(mpi::n_procs);
+      for (int i = 0; i < mpi::n_procs; i++) {
+        all_position[i].resize(random_ray::n_source_regions);
+      }
+
+      // Copy master rank data into gathered vector for convenience
+      all_position[0] = random_ray::position;
+
+      // Receive all data into gather vector
+      for (int i = 1; i < mpi::n_procs; i++) {
+        MPI_Recv(all_position[i].data(), random_ray::n_source_regions * 3,
+          MPI_DOUBLE, i, 0, mpi::intracomm, MPI_STATUS_IGNORE);
+      }
+
+      // Scan through gathered data and pick first valid cell posiiton
+      for (int sr = 0; sr < random_ray::n_source_regions; sr++) {
+        if (random_ray::position_recorded[sr] == 1) {
+          for (int i = 0; i < mpi::n_procs; i++) {
+            if (all_position[i][sr].x != 0.0 || all_position[i][sr].y != 0.0 ||
+                all_position[i][sr].z != 0.0) {
+              random_ray::position[sr] = all_position[i][sr];
+              break;
+            }
+          }
+        }
+      }
+    } else {
+      // Other ranks just send in their data
+      MPI_Send(random_ray::position.data(), random_ray::n_source_regions * 3,
+        MPI_DOUBLE, 0, 0, mpi::intracomm);
+    }
+  }
+
+  // For the rest of the source region data, we simply perform an all reduce,
+  // as these values will be needed on all ranks for transport during the
+  // next iteration.
+  MPI_Allreduce(MPI_IN_PLACE, random_ray::volume.data(),
+    random_ray::n_source_regions, MPI_DOUBLE, MPI_SUM, mpi::intracomm);
+
+  MPI_Allreduce(MPI_IN_PLACE, random_ray::was_hit.data(),
+    random_ray::n_source_regions, MPI_INT, MPI_SUM, mpi::intracomm);
+
+  MPI_Allreduce(MPI_IN_PLACE, random_ray::scalar_flux_new.data(),
+    random_ray::n_source_elements, MPI_FLOAT, MPI_SUM, mpi::intracomm);
+
+  simulation::time_bank_sendrecv.stop();
+#endif
+}
+
+int openmc_run_random_ray()
+{
+  // Initialize OpenMC general data structures
+  openmc_simulation_init();
+
+  // Validate that inputs meet requirements for random ray mode
+  if (mpi::master)
+    validate_random_ray_inputs();
+
+  // Intialize Cell (Flat Source Region) data structures
+  initialize_source_regions();
+
+  // Random ray mode does not have an inner loop over generations within a
+  // batch, so set the current gen to 1
+  simulation::current_gen = 1;
+
+  // There are no source sites in random ray mode, so be sure to disable to
+  // ensure we don't attempt to write source sites to statepoint
+  settings::source_write = false;
+
+  // If all source regions have been hit by a ray during transport, then we can
+  // skip the mapping process for future iterations.
+  bool mapped_all_tallies = false;
+
+  // Random ray eigenvalue
+  // TODO: Use OpenMC native eigenvalue?
```

Comment:
Hm, keeping it separate makes a lot of sense to me; the functions for updating it would be entirely different.

Reply:
I'll keep them separate and take out the TODO. I may revisit this down the line when getting statepoint support going to see if something reasonable can be done or not.